client_code/kompot/_persist.py

```diff
@@ -0,0 +1,248 @@
+            return self._row[name]
+        else:
+            return getattr(self._despatcher, name, None)
+
+    def __setattr__(self, name, value):
+        if name.startswith("_") or not self._initialised:
+            object.__setattr__(self, name, value)
+        else:
+            setattr(self._despatcher, name, value)
+
+    def _despatch(self, action):
+        result = call(self._server_functions[action], self.despatcher.serialize())
+        self.despatcher.clear()
+        return result
+
+    def get(self, *args, **kwargs):
+        "Fetch a data tables row and make it available to the parent object"
+        self._row = anvil.server.call(self._server_functions["get"], *args, **kwargs)
+        self._despatcher.clear()
+
+    def create(self):
+        """Create a new data tables row from the parent object"""
+        return self._despatch("create")
+
+    def save(self):
+        """Save changes made to the parent object to the data tables row"""
+        action = "update" if self._row else "create"
+        return self._despatch(action)
+
+    def delete(self):
+        """Delete the data tables row"""
+        return self._despatch("delete")
+
+
+def _get_value(self, key):
+    if not key.startswith("_"):
+        try:
+            return getattr(self._store, key)
+        except AttributeError:
+            raise AttributeError(
+                f"{self.__class__.__name__} instance has no store defined"
+            )
+
+
+def _set_value(self, key, value):
+    if key.startswith("_"):
+        object.__setattr__(self, key, value)
+    else:
+        try:
+            setattr(self._store, key, value)
+        except AttributeError:
+            raise AttributeError(
+                f"{self.__class__.__name__} instance has no store defined"
+            )
+
+
+def _get(self, *args, **kwargs):
+    return self._store.get(*args, **kwargs)
+
+
+def _create(self):
+    return self._store.create()
+
+
+def _save(self):
+    return self._store.save()
+
+
+def _delete(self):
+    return self._store.delete()
+
+
+def persisted_class(cls):
+    """A decorator for a class with a persistence mechanism"""
+    members = {
+        "__module__": cls.__module__,
+        "__getattr__": _get_value,
+        "__getitem__": _get_value,
+        "__setattr__": _set_value,
+        "__setitem__": _set_value,
+        "get": _get,
+        "create": _create,
+        "save": _save,
+        "delete": _delete,
+    }
+    class_members = {k: v for k, v in cls.__dict__.items() if k not in members}
+    members.update(class_members)
+    return type(cls.__name__, (object,), members)
+
+
+def _row_backed_constructor(self):
+    def init(self, row=None):
+        self.store = RowBackedStore(self.server_functions, self.linked_attributes, row)
+
+    return init
+
+
+def row_backed_class(cls):
+    """A decorator for a class persisted by a data tables row"""
+    _cls = persisted_class(cls)
+    setattr(_cls, "__init__", _row_backed_constructor)
```

Comment:
What am I missing here?
we're setting `__init__` to `_row_backed_constructor` which is a function that returns a function
Doesn't seem to make sense.

Reply:
yeah, this went around the houses a few times, so is a bit of a mess right now.